### PR TITLE
Refactor all (demo) runtime modules to use new storage

### PR DESCRIPTION
**Pull request by gavofyork**

### CODE REVIEW

File: demo/runtime/src/genesismap.rs [hunk 1]
```diff
@@ -80,40 +80,37 @@ impl GenesisConfig {
 	pub fn genesis_map(&self) -> HashMap<Vec<u8>, Vec<u8>> {
 		let wasm_runtime = include_bytes!("../wasm/genesis.wasm").to_vec();
 		vec![
-			(&session::SESSION_LENGTH[..], vec![].and(&self.session_length)),
-			(&session::VALIDATOR_COUNT[..], vec![].and(&(self.validators.len() as u32))),
+			(session::SessionLength::key(), vec![].and(&self.session_length)),
```
Comment (rphmeier) [inline demo/runtime/src/genesismap.rs]: 🎉  🎊

File: substrate/runtime-support/src/storage/mod.rs [hunk 2]
```diff
@@ -122,58 +122,77 @@ pub fn put_raw(key: &[u8], value: &[u8]) {
 	runtime_io::set_storage(&twox_128(key)[..], value)
 }
 
-struct RuntimeStorage;
+pub struct RuntimeStorage;
```
Comment (rphmeier) [inline substrate/runtime-support/src/storage/mod.rs]: doe it need to be pub? should have docs if so

File: substrate/runtime-support/src/storage/generator.rs [hunk 3]
```diff
@@ -48,46 +48,78 @@
 
 use codec;
 use rstd::vec::Vec;
+pub use rstd::borrow::Borrow;
```
Comment (rphmeier) [inline substrate/runtime-support/src/storage/generator.rs]: `#[doc(hidden)]`

File: substrate/runtime-support/src/storage/generator.rs [hunk 4]
```diff
@@ -48,46 +48,78 @@
...
 pub trait Storage {
+	/// true if the key exists in storage.
+	fn exists(&self, key: &[u8]) -> bool;
+
+	/// Load the bytes of a key from storage. Can panic if the type is incorrect.
+	fn get<T: codec::Slicable>(&self, key: &[u8]) -> Option<T>;
+
 	/// Load the bytes of a key from storage. Can panic if the type is incorrect.
-	fn load<T: codec::Slicable>(&self, key: &[u8]) -> Option<T>;
+	fn require<T: codec::Slicable>(&self, key: &[u8]) -> T { self.get(key).expect("Required values must be in storage") }
```
Comment (rphmeier) [inline substrate/runtime-support/src/storage/generator.rs]: doc string should specify that it panics if it's not there

File: substrate/runtime-support/src/storage/generator.rs [hunk 5]
```diff
@@ -48,46 +48,78 @@
...
 	/// Take a value from storage, deleting it after reading.
 	fn take<T: codec::Slicable>(&self, key: &[u8]) -> Option<T> {
-		let value = self.load(key);
+		let value = self.get(key);
 		self.kill(key);
 		value
 	}
+
+	/// Take a value from storage, deleting it after reading.
+	fn seize<T: codec::Slicable>(&self, key: &[u8]) -> T { self.take(key).expect("Required values must be in storage") }
```
Comment (rphmeier) [inline substrate/runtime-support/src/storage/generator.rs]: `take/seize` is confusing.

File: substrate/runtime-support/src/storage/generator.rs [hunk 6]
```diff
@@ -114,225 +146,119 @@ pub trait StorageList<T: codec::Slicable> {
...
 }
 
 #[macro_export]
 #[doc(hidden)]
 macro_rules! __storage_items_internal {
 	// generator for values.
-	(($($vis:tt)*) $name: ident: $key: expr => $ty:ty) => {
+	(($($vis:tt)*) ($get_fn:ident) ($gettype:ty) ($getter:ident) ($taker:ident) $name:ident : $key:expr => $ty:ty) => {
+		__storage_items_internal!{ ($($vis)*) () ($gettype) ($getter) ($taker) $name : $key => $ty }
+		pub fn $get_fn() -> $gettype { <$name as $crate::storage::generator::StorageValue<$ty>> :: get(&$crate::storage::RuntimeStorage) }
```
Comment (rphmeier) [inline substrate/runtime-support/src/storage/generator.rs]: tightly coupling the macro to `RuntimeStorage` (which is useless outside of runtime context) isn't helpful.

We have a `$name::get()` function anyway via the parent `storage` module.

File: substrate/runtime-support/src/storage/mod.rs (hunk 2 above)
Comment (gavofyork) [reply substrate/runtime-support/src/storage/mod.rs]: yup it's needed for the macro

File: substrate/runtime-support/src/storage/generator.rs (hunk 6 above)
Comment (gavofyork) [reply substrate/runtime-support/src/storage/generator.rs]: yeah but it's crappy to have half of the getters use the idiomatic `lower_case_function()` and half using `PascalCaseName::get()`. you need to guess which is which.

Comment (gavofyork) [reply substrate/runtime-support/src/storage/generator.rs]: if there's a better way of doing this, then happy to go with it, but i don't see it.

Comment (rphmeier) [reply substrate/runtime-support/src/storage/generator.rs]: fine for now, although I dislike the coupling.

**Pull request merged**